src/controllers/dynakube/version/updater.go

```diff
@@ -0,0 +1,108 @@
+package version
+
+import (
+	"fmt"
+
+	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/src/api/v1beta1"
+	"github.com/Dynatrace/dynatrace-operator/src/dockerconfig"
+	"github.com/Dynatrace/dynatrace-operator/src/dtclient"
+	"github.com/pkg/errors"
+)
+
+type versionStatusUpdater interface {
+	Name() string
+	Enabled() bool
+	Target() *dynatracev1beta1.VersionStatus
+	CustomImage() string
+	CustomVersion() string
+	AutoUpdate() bool
+	LatestImageInfo() (*dtclient.LatestImageInfo, error)
+	UseDefault(dockerCfg *dockerconfig.DockerConfig) error
+}
+
+func (reconciler *Reconciler) run(updater versionStatusUpdater, dockerCfg *dockerconfig.DockerConfig) error {
+	currentSource := reconciler.determineSource(updater)
+	var err error
+	defer func() {
+		if err == nil {
+			updater.Target().LastProbe = reconciler.timeProvider.Now()
+			updater.Target().Source = currentSource
+		}
+	}()
+
+	customImage := dtclient.ImageInfoFromUri(updater.CustomImage())
+	if customImage != nil {
+		err = updateVersionStatus(updater.Target(), customImage, reconciler.hashFunc, dockerCfg)
+		return err
+	}
+
+	if !updater.AutoUpdate() {
+		emptyVersionStatus := dynatracev1beta1.VersionStatus{}
+		previousSource := updater.Target().Source
+		if updater.Target() == nil || *updater.Target() == emptyVersionStatus {
+			log.Info(fmt.Sprintf("initial status update in progress for %s with no auto update", updater.Name()))
+		} else if previousSource == currentSource {
+			log.Info(fmt.Sprintf("status updated skipped for %s, due to no auto update", updater.Name()))
+			return nil
+		}
+	}
+
+	if reconciler.dynakube.FeaturePublicRegistry() {
+		var publicImage *dtclient.LatestImageInfo
+		publicImage, err = updater.LatestImageInfo()
+		if err != nil {
+			log.Info(fmt.Sprintf("could not get %s public image", updater.Name()))
+			return err
+		}
+		err = updateVersionStatus(updater.Target(), publicImage, reconciler.hashFunc, dockerCfg)
+		if err != nil {
+			log.Info(fmt.Sprintf("could not update %s status according to the public registry", updater.Name()))
+			return err
+		}
+		updater.Target().Version = updater.Target().ImageTag
+		return nil
+	}
+
+	err = updater.UseDefault(dockerCfg)
+	return err
+}
+
+func (reconciler *Reconciler) determineSource(updater versionStatusUpdater) dynatracev1beta1.VersionSource {
+	if updater.CustomImage() != "" {
+		return dynatracev1beta1.CustomImageVersionSource
+	}
+	if reconciler.dynakube.FeaturePublicRegistry() {
+		return dynatracev1beta1.PublicRegistryVersionSource
+	}
+	if updater.CustomVersion() != "" {
+		return dynatracev1beta1.CustomVersionVersionSource
+	}
+	return dynatracev1beta1.DefaultVersionSource
+}
+
+func updateVersionStatus(
+	target *dynatracev1beta1.VersionStatus,
+	image *dtclient.LatestImageInfo,
+	hashFunc ImageHashFunc,
+	dockerCfg *dockerconfig.DockerConfig,
+) error {
```

Comment:
Indenting looks weird to me.

Reply:
on github tabs look weird, in the actual IDE it looks fine